File: python/0002_add_two_numbers.py
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy = ListNode(-1)
        carry = 0
        cur = dummy
        while l1 or l2:
            if l1 and l2:
                sum = carry + l1.val + l2.val
                sum, carry = sum % 10, sum // 10
                cur.next = ListNode(sum)
                cur, l1, l2 = cur.next, l1.next, l2.next
            elif l1:
                sum = carry + l1.val
                sum, carry = sum % 10, sum // 10
                cur.next = ListNode(sum)
                cur, l1 = cur.next, l1.next
            else:
                sum = carry + l2.val
                sum, carry = sum % 10, sum // 10
                cur.next = ListNode(sum)
                cur, l2 = cur.next, l2.next
        if carry == 1:
            cur.next = ListNode(1)
        return dummy.next

File: go/0002_add_two_numbers.go
package main

// ListNode represents an element of linked list
type ListNode struct {
	Next *ListNode
	Val  int
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	cur := &ListNode{}
	head := cur

	sum, co := 0, 0
	for co != 0 || l1 != nil || l2 != nil {
		if l1 == nil && co == 0 {
			cur.Next = l2
			break
		}
		if l2 == nil && co == 0 {
			cur.Next = l1
			break
		}

		cur.Next = &ListNode{}
		cur = cur.Next

		sum = co
		if l1 != nil {
			sum += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			sum += l2.Val
			l2 = l2.Next
		}

		co = sum / 10
		cur.Val = sum % 10
	}
	return head.Next
}
